librad/src/net/discovery/mdns.rs

```diff
@@ -89,6 +151,106 @@ impl Discovery for Mdns {
     }
 }
 
+struct Timestamped<T> {
+    timestamp: Instant,
+    inner: T,
+}
+
+impl<T> PartialOrd for Timestamped<T> {
+    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+        Some(self.timestamp.cmp(&other.timestamp).reverse())
+    }
+}
+
+impl<T> Ord for Timestamped<T> {
+    fn cmp(&self, other: &Self) -> cmp::Ordering {
+        self.timestamp.cmp(&other.timestamp).reverse()
+    }
+}
+
+impl<T> PartialEq for Timestamped<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.timestamp.eq(&other.timestamp)
+    }
+}
+
+impl<T> Eq for Timestamped<T> {}
+
+#[derive(Debug)]
+enum Discovered<T> {
+    Known,
+    New(T),
+    Refreshed(T),
+}
+
+struct Discoveries<K, V> {
+    all: FxHashMap<K, (Instant, V)>,
+    hip: BinaryHeap<Timestamped<K>>,
+    ttl: Duration,
+}
+
+impl<K, V> Discoveries<K, V>
+where
+    K: Clone + Eq + Hash,
+    V: Clone + PartialEq,
+{
+    pub fn new(ttl: Duration) -> Self {
+        Self {
+            all: Default::default(),
+            hip: Default::default(),
+            ttl,
+        }
+    }
+
+    pub fn insert(&mut self, k: K, v: V) -> Discovered<(K, V)> {
+        use std::collections::hash_map::Entry::*;
+        use Discovered::*;
+
+        let now = Instant::now();
+        self.evict(now - self.ttl);
```

Comment:
In any case you need some form of trigger, if I understand it correctly. What other trigger would you propose?

Reply:
Because otherwise we'd need to either spawn a thread, or the eviction is up to the caller (ie. that datastructure doesn't maintain any of its eviction guarantees).

Reply:
Right, I thought that might be the case and just wanted to make sure.